File: myapp/src/lmm/jwas.py
# import os
# os.chdir("../../..")
# os.getcwd()
from ..lib import *
from . import path as PATH

# for anaconda interpreter
import sys
if "conda" in sys.executable:
    print("It's Conda Python")
    from julia.api import Julia
    jl = Julia(compiled_modules=False)

# import Julia packages
from julia import JWAS
from julia import DataFrames
from julia import CSV

def call_JWAS():
    # take inputs
    dt_param = pd.read_csv(PATH.param_Julia, header=None)
    ARG = dict()
    ARG["data"] =   dt_param.iloc[0].values[0]
    ARG["ped"] =    dt_param.iloc[1].values[0]
    ARG["eq"] =     dt_param.iloc[2].values[0]
    ARG["cov"] =    dt_param.iloc[3].values[0]
    ARG["rdmstr"] = dt_param.iloc[4].values[0]
    ARG["rdmiid"] = dt_param.iloc[5].values[0]
    ARG["vgstr"] =  dt_param.iloc[6].values[0]
    ARG["vgiid"] =  dt_param.iloc[7].values[0]
    ARG["vgres"] =  dt_param.iloc[8].values[0]

    PATH_OUT = "myapp/out/jwas_%s.csv"
    # read data
    julia_dt = CSV.read(ARG["data"], DataFrames.DataFrame,
                        missingstrings=["0", ""])
    ped = JWAS.get_pedigree(ARG["ped"], header=True,
                            separator=",", missingstrings=["0", ""])

    # build pedigree
    id_ped = pd.Series(ped.IDs).str.\
                replace("[a-zA-Z]", "").\
                astype(int)
    id_int = np.array([int(ID) for ID in id_ped])
    order_ped = np.argsort(id_int)
    id_order = id_int[order_ped]

    # compute inversed A
    Ai = np.array(JWAS.PedModule.AInverse(ped))
    A = np.linalg.inv(Ai).round(2)
    dt_ped = pd.DataFrame(A).iloc[order_ped, order_ped]
    dt_ped.columns = np.array(ped.IDs)[order_ped]
    dt_ped.to_csv(PATH_OUT % "A", index=False)

    # build model equation
    model = JWAS.build_model(ARG["eq"], np.array(
        pd.read_csv(ARG["vgres"]).iloc[:, 1:])[0][0])

    # factor or covariates
    if ARG['cov'] == ARG['cov']:
        # is not nan field
        JWAS.set_covariate(model, ARG["cov"])

    # fixed or random
    if ARG['rdmstr'] == ARG['rdmstr']:
        # is not nan field
        JWAS.set_random(model, ARG["rdmstr"], ped, np.array(
            pd.read_csv(ARG["vgstr"]).iloc[:, 1:]))

    if ARG['rdmiid'] == ARG['rdmiid']:
        # is not nan field
        # JWAS.set_random(model, ARG["rdmiid"], ARG["vu"])
        JWAS.set_random(model, ARG["rdmiid"], np.array(
            pd.read_csv(ARG["vgiid"]).iloc[:, 1:]))

    # solve
    out = JWAS.solve(model, julia_dt)
    LHS = out[2]
    RHS = out[3]
    sol = pd.DataFrame(dict({"terms": out[0],
                            "values": np.array(np.linalg.pinv(LHS) @ RHS).
                            reshape(-1)}))

    # organize output
    ls_terms = re.split(r"\s*=\s*", ARG["eq"])
    name_y = ls_terms[0]

    # get random index
    ls_rdm = []
    if ARG["rdmstr"] == ARG["rdmstr"]:
        ls_rdm += re.split(r"\s", ARG["rdmstr"])
    if ARG["rdmiid"] == ARG["rdmiid"]:
        ls_rdm += re.split(r"\s", ARG["rdmiid"])

    idx_rdm = []
    for term in ls_rdm:
        idx_rdm += [i for i in range(len(sol)) if term in sol.iloc[i, 0]]

    # make solution df
    def format_terms(array):
        arr = [re.sub("%s:" % name_y, "", item) for item in array]
        arr = [re.sub("intercept:intercept", "*intercept:0", item) for item in arr]
        arr = [re.sub(":", "_", item) for item in arr]
        return np.array(arr).astype(str)

    dt_fix = sol.drop(idx_rdm)
    dt_fix.iloc[:, 0] = format_terms(dt_fix.iloc[:, 0])
    if len(dt_fix) != 0:
        dt_fix.loc[:, "isFixed"] = 1

    dt_rdm = sol.loc[idx_rdm, :]
    dt_rdm.iloc[:, 0] = format_terms(dt_rdm.iloc[:, 0])
    if len(dt_rdm) != 0:
        dt_rdm.loc[:, "isFixed"] = 0

    dt_sol = pd.concat([dt_fix, dt_rdm])
    dt_sol.columns = ["terms", "effects", "isFixed"]

    # get the right order
    terms_origin = format_terms(sol.iloc[:, 0])
    dt_split = dt_sol["terms"].str.split("_", expand=True)
    dt_sol.loc[:, "terms_name"] = dt_split.iloc[:, 0]
    dt_sol.loc[:, "terms_int"] = dt_split.iloc[:, 1].\
                                    str.\
                                    replace("[a-zA-Z]", "0").\
                                    astype(int)

    dt_sol = dt_sol.sort_values(by=["isFixed", "terms_name", "terms_int"],
                                ascending=[False, True, True])
    terms_order = dt_sol.terms.values
    order = [list(terms_origin).index(term) for term in terms_order]

    # ourput solution
    dt_sol.loc[:, "terms"] = dt_sol.terms.str.replace(r".*intercept.*", "intercept")
    dt_sol.iloc[:, :3].to_csv(PATH_OUT % "sol", index=False)
    ls_isFixed = dt_sol.isFixed.values.astype(bool)

    # make incidence matrix
    dt_inc = pd.DataFrame(np.array(out[1])[:, order])
    try:
        dt_fix = dt_inc.loc[:, ls_isFixed]
        dt_fix.columns = dt_sol.terms[ls_isFixed].values
        dt_fix.to_csv(PATH_OUT % "X", index=False)
    except:
        pd.DataFrame().to_csv(PATH_OUT % "X", index=False)

    try:
        dt_rdm = dt_inc.loc[:, ~ls_isFixed]
        dt_rdm.columns = dt_sol.terms[~ls_isFixed].values
        dt_rdm.to_csv(PATH_OUT % "Z", index=False)
    except:
        pd.DataFrame().to_csv(PATH_OUT % "Z", index=False)

    # LHS, RHS
    pd.DataFrame(np.array(out[2])[order, :][:, order]).to_csv(PATH_OUT % "lhs", index=False)
    pd.DataFrame(np.array(out[3])[order, :]).to_csv(PATH_OUT % "rhs", index=False)

File: myapp/test/jwas.jl
# Imports
# import Pkg
Pkg.add(["JWAS", "DataFrames", "CSV", "InvertedIndices"])
using JWAS, DataFrames, CSV, InvertedIndices

# Step 1: Take inputs
df_param = CSV.read("myapp/out/param.csv", DataFrame, header=false, delim=",")
ARG = Dict()
ARG["data"] = df_param[1, 1]
ARG["pedigree"] = df_param[2, 1]
ARG["eq"] = df_param[3, 1]
ARG["cov"] = df_param[4, 1]
ARG["rdmstr"] = df_param[5, 1]
ARG["rdmiid"] = df_param[6, 1]
ARG["vgstr"] = df_param[7, 1]
ARG["vgiid"] = df_param[8, 1]
ARG["vgres"] = df_param[9, 1]

# Step 2: Read data
phenotypes = CSV.read(ARG["data"],
    DataFrame,
    delim=',',
    header=true,
    missingstrings=["NA"])

# Step 2-2: Build Pedigree
ped = get_pedigree(ARG["pedigree"], header=true, separator=",")

id_int = map(x->parse(Int8, x), ped.IDs)
order_ped = sortperm(id_int)

# compute inversed A
Ai = JWAS.PedModule.AInverse(ped)
# from sparse to dense and get A
AiD = Matrix(Ai)
AD = inv(AiD)
CSV.write("myapp/out/jwas_ped.csv", DataFrame(AD)[order_ped, order_ped], header=false)

# Step 3: Build Model Equationcovs
model = build_model(ARG["eq"], ARG["ve"]); # set residual var,
# n-trait model
# (model1; model2, nxn matrix)

# Step 4: Set Factors or Covariates
if !(ARG["cov"] === missing)
    set_covariate(model, ARG["cov"]);
end

# Step 5: Set Random or Fixed Effects
if !(ARG["rdmstr"] === missing)
    set_random(model, ARG["rdmstr"], ped); # set random var (ped, matrix V)
end
if !(ARG["rdmiid"] === missing)
    dt_vu = CSV.read(ARG["vg"], DataFrame)
    set_random(model, ARG["rdmiid"], Matrix(dt_vu)); # set random var
end


# Step 6: Solve Mixed Model Equations
sol = solve(model, phenotypes, solver="Gibbs");
out = solve(model, phenotypes);

# Step 7: Organize outputs
# get response variables
ls_terms = split(ARG["eq"], r"\s*=\s*")
name_y = ls_terms[1]

# get random index 
ls_rdm = String[]
if !(ARG["rdmstr"] === missing)
    ls_rdm = vcat(ls_rdm, split(ARG["rdmstr"], r"\s"))
end
if !(ARG["rdmiid"] === missing)
    ls_rdmiid = vcat(ls_rdm, split(ARG["rdmiid"], r"\s"))
end

idx_rdm = Int8[]
for rdm in ls_rdm
    idx = findall(x -> occursin(rdm, x), sol[:, 1])
    append!(idx_rdm, idx)
end

# Step 8.1 Make solution DF
df_fix = DataFrame()
df_fix[:, :terms] = replace.(sol[Not(idx_rdm), 1], "$name_y:"=>"")
df_fix[:, :terms] = replace.(df_fix.terms, "intercept:intercept"=>"*intercept:0") # make it sorted the first
df_fix[:, :terms] = replace.(df_fix.terms, ":"=>"_")
df_fix[:, :effects] = sol[Not(idx_rdm), 2]
df_fix[:, :isFixed] .= 1

df_rdm = DataFrame()
df_rdm[:, :terms] = replace.(sol[idx_rdm, 1], "$name_y:"=>"")
df_rdm[:, :terms] = replace.(df_rdm.terms, ":"=>"_")
df_rdm[:, :effects] = sol[idx_rdm, 2]
df_rdm[:, :isFixed] .= 0
df_sol = vcat(df_fix, df_rdm)

# Step 8.1.2: Get right order
dt_terms = split.(df_sol.terms, '_')
foreach(enumerate([:Y1, :Y2])) do (i, n)
    df_sol[!, n] = getindex.(dt_terms, i)
end
df_sol[!, :Y2] = map(x->parse(Float64, x), df_sol.Y2)
idx_sort_all = sortperm(df_sol, [:isFixed, :Y1, :Y2], rev=(true, false, false))
idx_sort_X = sortperm(df_sol[[x in 1 for x in df_sol[!, :isFixed]] ,:], [:Y1, :Y2])
idx_sort_Z = sortperm(df_sol[[x in 0 for x in df_sol[!, :isFixed]] ,:], [:Y1, :Y2])


CSV.write("myapp/out/jwas_sol.csv", df_sol[idx_sort_all, :])

# Step 8.2 Make design DF
df_idc_rdm = DataFrame(Matrix(out[2])[:, idx_rdm])
if size(df_idc_rdm) != (0, 0)
    DataFrames.rename!(df_idc_rdm, df_rdm.terms)
    CSV.write("myapp/out/jwas_Z.csv", df_idc_rdm[:, idx_sort_Z])
else
    CSV.write("myapp/out/jwas_Z.csv", DataFrame(Empty_Data = []))
end

df_idc_fix = DataFrame(Matrix(out[2])[:, Not(idx_rdm)])
if size(df_idc_fix) != (0, 0)
    DataFrames.rename!(df_idc_fix, df_fix.terms)
    CSV.write("myapp/out/jwas_X.csv", df_idc_fix[:, idx_sort_X])
else
    CSV.write("myapp/out/jwas_X.csv", DataFrame(Empty_Data = []))
end

# Step 8.3 LHS and RHS
CSV.write("myapp/out/jwas_lhs.csv", DataFrame(Matrix(out[3]))[idx_sort_all, idx_sort_all])
CSV.write("myapp/out/jwas_rhs.csv", DataFrame(Matrix(out[4]))[idx_sort_all, :])

# NOTE
# ids, Ai, inbred = get_info(ped,Ai=true)
# set_random(model, "Animal Dam", Vinv=Ai, names=ids); # set random var
# A=[ids abs.(round.(inv(Matrix(Ai)),digits=2))]
# print(versioninfo())
